src/hera/workflows/_context.py

```diff
@@ -76,27 +76,29 @@ def add_sub_node(self, node: Union[SubNodeMixin, TTemplate]) -> None:
         if not pieces:
             return
 
-        try:
-            # here, we are trying to add a node to the last piece of context in the hopes that it is a subbable
-            pieces[-1]._add_sub(node)
-        except InvalidType:
-            # if the above fails, it means the user invoked a decorated function e.g. `@script`. Hence,
-            # the object needs to be added as a template to the piece of context at [-1]. This will be the case for
-            # DAGs and Steps
-            pieces[-1]._add_sub(node.template)  # type: ignore
-
-        # when the above does not raise an exception, it means the user invoked a decorated function e.g. `@script`
+        # for particular types, the user invoked a decorated function e.g. `@script`
         # inside a proper context. Here, we add the object to the overall workflow context, directly as a template,
         # in case it is not found (based on the name). This helps users save on the number of templates that are
         # added when using an object that is a `Script`
```

Comment:
We can simplify/correct the comment
```suggestion
        # When the user invokes a decorated function e.g. `@script inside a sub-context (dag/steps),
        # we also add the step/task's template to the overall workflow context, if it is not already added.
```